### PR TITLE
map (mapbox) mvp app extension

### PR DESCRIPTION
**Pull request by sunil2342**

### CODE REVIEW

File: ui/src/mixins/Mapbox/controls.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+export default (instance, mapboxgl, position = "top-left") => {
+  // valid values for positions
+  const validPositions = [
```
Comment (sonokoh) [inline ui/src/mixins/Mapbox/controls.js:3]: Lets declare constants like these outside function so we don't declare them on every single func call

File: ui/src/init/index.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+const config = require("./_config");
```
Comment (sonokoh) [inline ui/src/init/index.js:1]: I assume you are trying to dynamically import mapbox library via CDN. Any reason you avoided importing it as npm. I think this creates unnecessary code complexity (I see this is the reason you created find.js). Can we just use npm instead? 

Even If we find out we want to use cdn, I believe we can modify webpack chain via Quasar app extension capability(https://quasar.dev/app-extensions/tips-and-tricks/chain-webpack) and use CDN loader.

File: ui/src/directives/maps/controls.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import Vue from "vue";
```
Comment (sonokoh) [inline ui/src/directives/maps/controls.js:1]: are you using this directive anywhere? I couldnt find the usage

File: ui/src/components/MapBox.vue [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  mixins: [MapBox],
+  data() {
+    return {
+      markers: [],
+    };
+  },
+  created() {
+    const { mapboxgl } = window;
+    // marker events (setup marker if 'set-marker' is fired)
+    $bus.$on("set-marker", (settings) => {
```
Comment (sonokoh) [inline ui/src/components/MapBox.vue:21]: So in general global bus usage is discouraged by Vue style guide (https://vuejs.org/v2/style-guide/#Non-flux-state-management-use-with-caution). Usually there's always a better or recommended way to accomplish the same. 

In a simple component extension like this I believe we can stay away from it. 
For example, in this case, we can just let QMapBox accept `:markers` as prop and render QMarker in its template.

If we want to allow users to override such default behavior, we can use slotScope and provide them marker object (https://vuejs.org/v2/guide/components-slots.html#Scoped-Slots).

**Review by sonokoh — COMMENTED** [review]
Thanks for the PR! I think this is a great start. Pls let me know if you have any questions or opinions.

I also found out that there actually is vue component project for mapbox that seems decent. https://github.com/soal/vue-mapbox. Can you check?

**Comment by sonokoh** [comment]
Hi, when you make changes per PR feedbacks, can you not close the PR and rather update it by committing and pushing changes to the same branch? That way it's easier to keep track of our comments and identify files that has been updated by your changes.
I'm reopening this PR, but let me know if you have any objection. When you are ready for another review, pls send me another review request. Thanks!

**Comment by sunil2342** [comment]
@sonokoh I am using same PR there were some feedback left but when I commit may be it will again ask to revisit to check feedback so by the time I was fixing all feedback I thought lets close PR by the time

**Comment by sonokoh** [comment]
@sunil2342 Ok! Don't worry about pushes, I don't think I get notified them. Keep it open, and just push re-review button when it is ready.

File: ui/dev/src/pages/Test1.vue [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+</template>
+<script>
+export default {
+  name: "test1",
+  data() {
+    return {
+      config: {
+        mapbox: {
+          controls: "top-right",
+          style: "mapbox://styles/softraw/cki8h8ft60nkq19s73qmwby0o"
```
Comment (sonokoh) [inline ui/dev/src/pages/Test1.vue:32]: Again, this style did not work for me so I had to use mapbox://sprites/mapbox/streets-v8. Markers seems to be rendered in weird positions (rendered outside the default screen range, and they slides the map when I scale up/down )
If you have some instructions to test properly, please let me know.

File: ui/src/components/Marker.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    },
+  },
+  data() {
+    return {
+      marker: undefined,
+    };
+  },
+  inject: ["map", "mode"],
+  mounted() {
+    if (this.mode === "mapbox") {
```
Comment (sonokoh) [inline ui/src/components/Marker.js:58]: Just remember, mounted is called only once. Shouldn't we cleanup/rerender when marker objects change?

File: ui/src/index.sass [hunk 7]
```diff
@@ -0,0 +1,5 @@
+// Variables
+@import 'quasar/src/css/variables.sass'
+@import './css/mapbox-gl.scss'
```
Comment (sonokoh) [inline ui/src/index.sass:3]: I received error saying @import './css/mapbox-gl.scss' cannot be found. Did you forget to commit it?

File: ui/src/init/_config.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (sonokoh) [inline ui/src/init/_config.json:1]: Do we still need this file?

**Review by sonokoh — COMMENTED** [review]
Thanks for all the fixes, the project quite improved.
So, it seems you decided not to integrate vue-mapbox. Would you let me know why you made the decision? Thanks!

**Comment by sunil2342** [comment]
> Thanks for all the fixes, the project quite improved.
> So, it seems you decided not to integrate vue-mapbox. Would you let me know why you made the decision? Thanks!

The current production version of npm library of vue-mapbox is 0.4.1 which is outdated and last updated over 2yrs ago and development version is still getting worked on, also I faced some performance issues with maps on zoom-in and zoom-out. And in the end by using a package we will end up with a wrapper on already a wrapper of mapbox-gl and we always have to be depend on the package.

**Review by sonokoh — COMMENTED** [review]
@sunil2342 Thanks for researching and informing your decision! 
I attempted to re-review per your request but it seems many of my previous comments are not yet addressed. Would you revisit them and please either fix them or reply to the comments if you intend not to fix? Thanks!

File: ui/src/components/Marker.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      if (this.text) {
+        const popup = new mapboxgl.Popup({ ...this.popup_setup }).setText(
+          this.text
+        );
+        marker.setPopup(popup);
+      }
+
+      this.marker = marker;
+    } else if (this.mode === "gmaps") {
+      const { google } = window;
```
[the remark this answers is not in the document]
Comment (sunil2342) [reply ui/src/components/Marker.js]: We are using `@googlemaps/js-api-loader` library for google maps, and it sets `google` variable to window object, we can either de-structure from window object or use `google` object directly.

File: ui/dev/src/pages/Test1.vue (hunk 5 above)
Comment (sunil2342) [reply ui/dev/src/pages/Test1.vue:44]: Changed the default style to /streets-v11.  Issue was caused due to style not available publicly

**Review by sonokoh — APPROVED** [review]
Thanks for fixing them!

**Pull request merged**